File: day15/dueling_generators.cpp
#include <iostream>
using namespace std;

int main(int argc, char *argv[]) {
//    unsigned long gen[] = {65, 8921};
    unsigned long gen[] = {116, 299};
    const unsigned long FACT[] = {16807, 48271};
//    const int PAIRS = 5;
    const int PAIRS = 40000000;
    int cnt = 0;

    for (int i = 0; i < PAIRS; i++) {
        for (int j = 0; j < 2; j++)
            gen[j] = (gen[j] * FACT[j]) % 2147483647;

        if ((gen[0] & 0xffff) == (gen[1] & 0xffff))
            cnt++;
    }

    cout << "result1: " << cnt << endl;
}

File: day15/dueling_generators.m
function dueling_generators( initGen )
    tic
    gen = uint64(initGen);
    FACT = uint64([16807 48271]);
    PAIRS = 40000000;
    cnt = 0;
    MASK = hex2dec('ffff');
    
    for i = 1:PAIRS
        gen = mod(gen .* FACT, 2147483647);
        lowbits = bitand(gen, MASK);
        if lowbits(1) == lowbits(2)
            cnt = cnt + 1;
        end
        % fprintf('%10d %10d\n', gen(1), gen(2));
    end
    fprintf('result1: %d\n', cnt);
    toc
end

